File: backend/app/sentry.py
"""Sentry initialization for the FastAPI backend."""

from typing import Any

import sentry_sdk
from sentry_sdk.integrations.asyncpg import AsyncPGIntegration
from sentry_sdk.integrations.fastapi import FastApiIntegration

from app.config import get_settings


def init_sentry() -> None:
    settings = get_settings()
    dsn = getattr(settings, "SENTRY_DSN", None)
    if not dsn:
        return

    sentry_sdk.init(
        dsn=dsn,
        environment="production" if not settings.DEBUG else "development",
        integrations=[
            FastApiIntegration(transaction_style="endpoint"),
            AsyncPGIntegration(),
        ],
        # Performance: capture 20% of transactions
        traces_sample_rate=0.2 if not settings.DEBUG else 1.0,
        # PII sending disabled to protect user data
        send_default_pii=False,
        # Filter out health check noise
        before_send=_before_send,
    )


def _before_send(event: dict[str, Any], hint: dict[str, Any]) -> dict[str, Any] | None:
    # Don't send health check errors
    if event.get("request", {}).get("url", "").endswith("/health"):
        return None
    return event

File: frontend/src/lib/sentry.ts
import * as Sentry from '@sentry/react';

export function initSentry() {
  const dsn = import.meta.env.VITE_SENTRY_DSN;
  if (!dsn) return;

  Sentry.init({
    dsn,
    environment: import.meta.env.MODE,
    integrations: [
      Sentry.browserTracingIntegration(),
    ],
    // Performance: capture 20% of transactions in production
    tracesSampleRate: import.meta.env.PROD ? 0.2 : 1.0,
    // Session replay: 10% of sessions, 100% of sessions with errors
    replaysSessionSampleRate: 0.1,
    replaysOnErrorSampleRate: 0.5,
    // Filter out noise
    ignoreErrors: [
      'ResizeObserver loop',
      'Network request failed',
      'Load failed',
    ],
    beforeSend(event) {
      // Don't send in development
      if (import.meta.env.DEV) return null;
      return event;
    },
  });

  // Lazy-load session replay integration to reduce initial bundle (~70KB)
  Sentry.lazyLoadIntegration('replayIntegration').then((replayIntegration) => {
    Sentry.addIntegration(replayIntegration({ maskAllText: true, blockAllMedia: true }));
  });
}

// Re-export for use in components
export { Sentry };
